### PR TITLE
fixed #1384

### LINKED ISSUE #1384: Semicolon not allowed after function-local item
As seen in etaoins/arret@9e2989198556c8f0ef322cde0913d98299c65e05 `rust-analyzer` does not allow a semicolon after a function-local enum:

```rust
fn simple_function() {
    enum LocalEnum {
        One,
        Two,
    };
}
```

This is accepted by `rustc` but `rust-analyzer` complains with:

```rust
Syntax Error: expected item, found `;`
consider removing this semicolon
```

### PR DESCRIPTION
**Pull request by zbelial**
This PR fixed #1384 .

### CODE REVIEW

**Comment by matklad** [comment]
This will treat `;` as a part of enum variant list, which is not exactly right: in Rust, unlike C++, one should not add `;` after types. The code in the issue is accepted, because `;` is treated as an empty statement.

In other words, we want to get syntax tree like this:

```
  BLOCK@[21; 86)
    L_CURLY@[21; 22) "{"
    WHITESPACE@[22; 27) "\n    "
    ENUM_DEF@[27; 75)
      ENUM_KW@[27; 31) "enum"
      WHITESPACE@[31; 32) " "
      NAME@[32; 41)
        IDENT@[32; 41) "LocalEnum"
      WHITESPACE@[41; 42) " "
      ENUM_VARIANT_LIST@[42; 75)
        L_CURLY@[42; 43) "{"
        WHITESPACE@[43; 52) "\n        "
        ENUM_VARIANT@[52; 55)
          NAME@[52; 55)
            IDENT@[52; 55) "One"
        COMMA@[55; 56) ","
        WHITESPACE@[56; 65) "\n        "
        ENUM_VARIANT@[65; 68)
          NAME@[65; 68)
            IDENT@[65; 68) "Two"
        COMMA@[68; 69) ","
        WHITESPACE@[69; 74) "\n    "
        R_CURLY@[74; 75) "}"
    SEMI@[83; 84) ";"
    WHITESPACE@[84; 85) "\n"
    R_CURLY@[85; 86) "}"
```

and not like this:

```
  BLOCK@[21; 86)
    L_CURLY@[21; 22) "{"
    WHITESPACE@[22; 27) "\n    "
    ENUM_DEF@[27; 75)
      ENUM_KW@[27; 31) "enum"
      WHITESPACE@[31; 32) " "
      NAME@[32; 41)
        IDENT@[32; 41) "LocalEnum"
      WHITESPACE@[41; 42) " "
      ENUM_VARIANT_LIST@[42; 75)
        L_CURLY@[42; 43) "{"
        WHITESPACE@[43; 52) "\n        "
        ENUM_VARIANT@[52; 55)
          NAME@[52; 55)
            IDENT@[52; 55) "One"
        COMMA@[55; 56) ","
        WHITESPACE@[56; 65) "\n        "
        ENUM_VARIANT@[65; 68)
          NAME@[65; 68)
            IDENT@[65; 68) "Two"
        COMMA@[68; 69) ","
        WHITESPACE@[69; 74) "\n    "
        R_CURLY@[74; 75) "}"
        SEMI@[83; 84) ";"
    WHITESPACE@[84; 85) "\n"
    R_CURLY@[85; 86) "}"
```

I *think* `expr_block_contents` or `stmt` functions should be fixed. There's some logic in `expr_block_contents` for handling `;` but looks like it doesn't always work.

**Comment by zbelial** [comment]
Thanks for your review. I'll try again :)
Although I've read most parts of TRPL, I'm still a noob of Rust and the reason why I want to contribute to this project is that I think it might be an effective way to master Rust.  
I'll study more about Rust to avoid (or maybe just eliminate) making PR like this.

**Comment by zbelial** [comment]
Hi, after reading the according parts of the reference [here](https://doc.rust-lang.org/reference/index.html) and the code, I think the problem is in `items_without_modifiers`, at the end of which it checks whether the current token is semi or not.

In `expr_block_contents`, semi is handled, so I think removing the check at the end of `items_without_modifiers` will solve this issue. 

Am I right?

**Comment by matklad** [comment]
Yes, this looks like it!

We only should forbid a semicolon in modules.

**Comment by zbelial** [comment]
Ok, then I think I should make another PR and close this one.

File: crates/ra_syntax/tests/data/parser/err/0003_C++_semicolon.txt [hunk 1]
```diff
@@ -35,5 +35,4 @@ SOURCE_FILE@[0; 40)
       R_CURLY@[38; 39) "}"
   ERROR@[39; 40)
     SEMI@[39; 40) ";"
-error 39: expected item, found `;`
-consider removing this semicolon
```
Comment (matklad) [inline crates/ra_syntax/tests/data/parser/err/0003_C++_semicolon.txt]: In this case, this error is correct: on the top-level, semicolon is forbidden:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f80495a0fafa572b9b5f44c3fec8fc66

So we should still issue this specialized error message, but only if the item is on the top level.

Comment (zbelial) [reply crates/ra_syntax/tests/data/parser/err/0003_C++_semicolon.txt]: Ok, I'll fix this.

**Comment by matklad** [comment]
Changes look good to me now!

Could you squash them down to a single commit? The correct way is to do interractive rebase:

https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

but, in this case, the simpler approach might be to first `git reset --soft commit-hash` your branch to commit just *before* 
ca35f7a , commit changes again (this effectively squashes the commits into one) and `git push --force-with-lease` the branch.

**Comment by zbelial** [comment]
Ok, I'll do that. 
Thanks very much for you reviews and feedbacks.

**Comment by matklad** [comment]
Everything is perfect now, thanks!

bors r+

**Pull request merged**